src/systems/physics-system.js

```diff
@@ -0,0 +1,101 @@
+import { World } from "three-ammo";
+import { TYPE } from "three-ammo/constants";
+
+const WORLD_CONFIG = {
+  debugDrawMode: THREE.AmmoDebugConstants.DrawWireframe,
+  gravity: { x: 0, y: -9.8, z: 0 }
+};
+
+export class PhysicsSystem {
+  constructor(scene) {
+    const Ammo = require("ammo.js/builds/ammo.wasm.js");
+    const AmmoWasm = require("ammo.js/builds/ammo.wasm.wasm");
+    const AmmoModule = Ammo.bind(undefined, {
+      locateFile(path) {
+        if (path.endsWith(".wasm")) {
+          return AmmoWasm;
+        }
+        return path;
+      }
+    });
+
+    this.bodies = [];
+    this.bodyHelpers = [];
+    this.shapeHelpers = [];
+
+    this.debugRequested = false;
+    this.debugEnabled = false;
+    this.scene = scene;
+    this.stepDuration = 0;
+
+    AmmoModule().then(() => {
+      this.world = new World(WORLD_CONFIG);
+    });
+  }
+
+  setDebug(debug) {
+    this.debugRequested = debug;
+  }
+
+  tick(dt) {
+    if (this.world) {
+      if (this.debugRequested !== this.debugEnabled) {
+        this.debugEnabled = this.debugRequested;
+        if (this.debugEnabled) {
+          this.world.getDebugDrawer(this.scene).enable();
+        } else {
+          this.world.getDebugDrawer(this.scene).disable();
+        }
+      }
+
+      while (this.bodyHelpers.length > 0) {
+        this.bodyHelpers.shift().init2();
+      }
+      while (this.shapeHelpers.length > 0) {
+        this.shapeHelpers.shift().init2();
+      }
+
+      for (let i = 0; i < this.bodies.length; i++) {
+        this.bodies[i].updateShapes();
+        if (this.bodies[i].type !== TYPE.DYNAMIC) {
+          this.bodies[i].syncToPhysics();
+        }
+      }
+      const time = performance.now();
+      this.world.step(dt / 1000);
+      this.stepDuration = performance.now() - time;
+      for (let i = 0; i < this.bodies.length; i++) {
+        if (this.bodies[i].type === TYPE.DYNAMIC) {
+          this.bodies[i].syncFromPhysics();
+        }
+      }
+    }
+  }
+
+  addBody(body) {
+    this.bodies.push(body);
+  }
+
+  removeBody(body) {
+    const idx = this.bodies.indexOf(body);
+    if (idx !== -1) {
+      this.bodies.splice(idx, 1);
+    }
+  }
```

Comment:
It's not clear to me whether the order of the bodies is important to the operation of physics -- if it's not then you could use a set and this would be nicer.

Reply:
The order doesn't matter but you can't iterate a set without generating garbage because of iterator instantiation, so I tend to avoid using sets (and maps) when I have to iterate over the data frequently.

Reply:
I see, that makes sense, I forgot about that.